src/eos-update-server.c

```diff
@@ -0,0 +1,863 @@
+/* -*- mode: C; c-file-style: "gnu"; indent-tabs-mode: nil; -*-
+ *
+ * Copyright 2016 Kinvolk GmbH
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the
+ * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
+ * Boston, MA 02111-1307, USA.
+ *
+ * Author: Krzesimir Nowak <krzesimir@kinvolk.io>
+ */
+
+#include "eos-refcounted.h"
+#include "eos-util.h"
+
+#include <gio/gio.h>
+#include <glib-object.h>
+#include <glib.h>
+
+#include <libsoup/soup.h>
+
+#include <ostree.h>
+
+#include <systemd/sd-daemon.h>
+
+#include <errno.h>
+#include <stdio.h>
+#include <string.h>
+
+typedef struct Options Options;
+
+struct Options
+{
+  guint16 local_port;
+  gint timeout_seconds;
+  gchar *served_remote;
+};
+
+#define OPTIONS_CLEARED { 0u, 0, NULL }
+
+static gboolean
+local_port_goption (const gchar *option_name,
+                    const gchar *value,
+                    gpointer options_ptr,
+                    GError **error)
+{
+  guint64 number;
+  const gchar *endptr = NULL;
+  int saved_errno;
+  Options* options = options_ptr;
+
+  if (g_strcmp0 (option_name, "--local-port") != 0 && g_strcmp0 (option_name, "-p") != 0)
+    {
+      g_set_error (error, G_OPTION_ERROR, G_OPTION_ERROR_FAILED,
+                   "Wrong option %s to parse, should not happen", option_name);
+      return FALSE;
+    }
```

Comment:
Might be worth splitting this check out into a separate function that you can use like this:

    check_option_is(option_name, "--local-port", "-p", error)